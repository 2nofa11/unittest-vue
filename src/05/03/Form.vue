<script setup lang="ts">
defineProps<{ name: string }>();

const emit = defineEmits<{ (e: "submit", event: Event): void }>();
const handleSubmit = (event: Event) => {
  emit("submit", event);
};
</script>

<template>
  <form @submit.prevent="handleSubmit">
    <h2>アカウント情報</h2>
    <p>{{ name }}</p>
    <div>
      <button>編集する</button>
    </div>
  </form>
</template>
