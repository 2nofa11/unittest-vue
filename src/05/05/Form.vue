<script setup lang="ts">
import { ref } from "vue";
import Agreement from "./Agreement.vue";
import InputAccount from "./InputAccount.vue";

const email = ref("");
const password = ref("");
const checked = ref(false);
</script>

<template>
  <form @submit.prevent="handleSubmit" data-test="signup-form">
    <h2>新規アカウント登録</h2>
    <InputAccount v-model:email="email" v-model:password="password" />
    <Agreement v-model="checked" data-test="agreement-component" />
    <div>
      <button :disabled="!checked" data-test="submit-button">
        サインアップ
      </button>
    </div>
  </form>
</template>
