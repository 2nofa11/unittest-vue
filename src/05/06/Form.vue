<script setup lang="ts">
import { ref, computed } from "vue";
import ContactNumber from "./ContactNumber.vue";
import DeliveryAddress from "./DeliveryAddress.vue";
import PastDeliveryAddress from "./PastDeliveryAddress.vue";
import RegisterDeliveryAddress from "./RegisterDeliveryAddress.vue";

export interface AddressOption {
  id: string;
  value: string;
  children: string;
}

const props = defineProps<{
  deliveryAddresses?: AddressOption[];
}>();

const emit = defineEmits<{
  (e: "submit", values: Record<string, any>): void;
}>();

const phoneNumber = ref("");
const name = ref("");
const registerNewAddress = ref<boolean | undefined>(undefined); // Initially undefined
const selectedPastAddressValue = ref("");
const postalCode = ref("");
const prefecture = ref("");
const municipalities = ref("");
const streetAddress = ref("");

const hasPastAddresses = computed(
  () => (props.deliveryAddresses?.length || 0) > 0
);

const isPastAddressDisabled = computed(
  () => registerNewAddress.value === undefined
);

const handleFormSubmit = (event: Event) => {
  const formData = new FormData(event.target as HTMLFormElement);
  const values: { [k: string]: unknown } = {};
  formData.forEach((value, key) => (values[key] = value));
  emit("submit", values); // submit イベントを発行
};
</script>

<template>
  <form @submit.prevent="handleFormSubmit" data-test="form">
    <h2 data-test="form-heading">お届け先情報の入力</h2>

    <ContactNumber
      v-model:phoneNumber="phoneNumber"
      v-model:name="name"
      data-test="contact-number-component"
    />

    <template v-if="hasPastAddresses">
      <RegisterDeliveryAddress
        v-model="registerNewAddress"
        data-test="register-delivery-address-component"
      />
      <DeliveryAddress
        v-if="registerNewAddress === true"
        title="新しいお届け先"
        v-model:postalCode="postalCode"
        v-model:prefecture="prefecture"
        v-model:municipalities="municipalities"
        v-model:streetAddress="streetAddress"
        data-test="new-delivery-address-component"
      />
      <PastDeliveryAddress
        v-else
        :disabled="isPastAddressDisabled"
        :options="props.deliveryAddresses || []"
        v-model="selectedPastAddressValue"
        data-test="past-delivery-address-component"
      />
    </template>
    <template v-else>
      <DeliveryAddress
        v-model:postalCode="postalCode"
        v-model:prefecture="prefecture"
        v-model:municipalities="municipalities"
        v-model:streetAddress="streetAddress"
        data-test="default-delivery-address-component"
      />
    </template>

    <hr />
    <div>
      <button type="submit" data-test="submit-button">
        注文内容の確認へ進む
      </button>
    </div>
  </form>
</template>
