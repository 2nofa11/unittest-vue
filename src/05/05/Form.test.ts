import { mount } from "@vue/test-utils";
import Form from "./Form.vue";

describe("Form.vue", () => {
  test("主要エリアが表示されている", () => {
    const wrapper = mount(Form);
    expect(wrapper.find("h2").text()).toBe("新規アカウント登録");
    expect(wrapper.find('[data-test="input-account-legend"]').text()).toBe(
      "アカウント情報の入力"
    );
    expect(wrapper.find('[data-test="agreement-legend"]').text()).toBe(
      "利用規約の同意"
    );
    expect(wrapper.find('[data-test="submit-button"]').text()).toBe(
      "サインアップ"
    );
  });

  test("初期状態では「サインアップ」ボタンは非活性", () => {
    const wrapper = mount(Form);
    const button = wrapper.find('[data-test="submit-button"]');
    expect(button.attributes("disabled")).toBeDefined(); // disabled 属性が存在する
  });

  test("「利用規約の同意」チェックボックスをチェックすると「サインアップ」ボタンは活性化", async () => {
    const wrapper = mount(Form);
    const checkbox = wrapper.find('[data-test="agreement-checkbox"]');
    const button = wrapper.find('[data-test="submit-button"]');

    expect(button.attributes("disabled")).toBeDefined(); // 初期状態は非活性
    await checkbox.setValue(true);
    expect(button.attributes("disabled")).toBeUndefined(); // 活性化 (disabled 属性がなくなる)
  });

  test("Snapshot: 新規アカウント登録フォームが表示される", () => {
    const wrapper = mount(Form);
    expect(wrapper.html()).toMatchSnapshot();
  });
});
