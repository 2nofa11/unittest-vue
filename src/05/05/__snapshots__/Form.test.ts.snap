// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form.vue Snapshot: 新規アカウント登録フォームが表示される 1`] = `
"<form data-test="signup-form">
  <h2>新規アカウント登録</h2>
  <fieldset data-test="input-account-fieldset">
    <legend data-test="input-account-legend">アカウント情報の入力</legend>
    <div><label> メールアドレス <input type="text" placeholder="example@test.com" data-test="input-email"></label></div>
    <div><label> パスワード <input type="password" placeholder="8文字以上で入力" data-test="input-password"></label></div>
  </fieldset>
  <fieldset data-test="agreement-component">
    <legend data-test="agreement-legend">利用規約の同意</legend><label><input type="checkbox" data-test="agreement-checkbox"> 当サービスの<a href="/terms">利用規約</a>を確認し、これに同意します </label>
  </fieldset>
  <div><button disabled="" data-test="submit-button"> サインアップ </button></div>
</form>"
`;
